File: java/graph/MatrixBFS.java
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.HashSet;
import java.util.Set;
import java.lang.Math;

public class MatrixBFS {

    //Matrix (2D Grid)
    int[][] grid = {{0, 0, 0, 0},
                    {1, 1, 0, 0},
                    {0, 0, 0, 1},
                    {0, 1, 0, 0}};

    // Shortest path from top left to bottom right
    public int bfs(int[][] grid) {
        int ROWS = grid.length;
        int COLS = grid[0].length;
        Set<Integer[]> visit = new HashSet<Integer[]>();
        Deque<Integer[]> queue = new ArrayDeque<>();

        queue.add(new Integer[2]);
        visit.add(new Integer[2]);

        int length = 0;
        while (!queue.isEmpty()) {
            for (int i = 0; i < queue.size(); i++) {
                Integer pair[] = queue.poll();
                int r = pair[0], c = pair[1];
                if (r == ROWS - 1 && c == COLS - 1) {
                    return length;
                }    
                // We can directly build the four neighbors
                Integer[][] neighbors = {{r, c + 1}, {r, c - 1}, {r + 1, c}, {r - 1, c}};
                for (int j = 0; j < 4; j++) {
                    int newR = neighbors[j][0], newC = neighbors[j][1];
                    if (Math.min(newR, newC) < 0 || newR == ROWS || newC == COLS
                    || visit.contains(neighbors[j]) || grid[newR][newC] == 1) {
                        continue;
                    }
                    queue.add(neighbors[j]);
                    visit.add(neighbors[j]);
                }
            }
            length++;
        }
        return length; // This should never be called
    }
}

File: swift/graph/matrixBfs.swift
// Matrix (2D Grid)
let grid = [[0, 0, 0, 0],
            [1, 1, 0, 0],
            [0, 0, 0, 1],
            [0, 1, 0, 0]]

// Shortest path from top left to bottom right
func bfs(_ grid: [[Int]]) -> Int {
  let ROWS = grid.count, COLS = grid[0].count
  var visit = Array(repeating: Array(repeating: 0, count: COLS), count: ROWS)
  var queue: [(Int, Int)] = []
  queue.append((0, 0))
  visit[0][0] = 1
  
  var length = 0
  while queue.count > 0 {
    for _ in 0..<queue.count {
      let (r, c) = queue.removeFirst()
      if r == ROWS - 1 && c == COLS - 1 {
        return length
      }

      let neighbors = [[0, 1], [0, -1], [1, 0], [-1, 0]]
      for neighbor in neighbors {
        let dr = neighbor[0], dc = neighbor[1]
        if (min(r + dr, c + dc) < 0 || 
           r + dr == ROWS || c + dc == COLS ||
           visit[r + dr][c + dc] == 1 || grid[r + dr][c + dc] == 1) {
           continue
        }
        queue.append((r + dr, c + dc))
        visit[r + dr][c + dc] = 1
      }
    }
    length += 1
  }
  return length // This should never be called
}